File: crypto/src/main/java/io/oversec/one/crypto/gpg/OpenKeychainConnector.java
package io.oversec.one.crypto.gpg;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.pm.ServiceInfo;
import android.net.Uri;
import android.os.IBinder;
import io.oversec.one.crypto.AppsReceiver;
import org.openintents.openpgp.IOpenPgpService2;
import org.openintents.openpgp.OpenPgpError;
import org.openintents.openpgp.util.OpenPgpApi;
import roboguice.util.Ln;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class OpenKeychainConnector {
    public static final String PACKAGE_NAME = "org.sufficientlysecure.keychain";
    //public static final String PACKAGE_NAME = "org.sufficientlysecure.keychain.debug";

    public final static int V_MIN = 39500;  //The minimum version we support, kind of arbitrary decision
    public final static int V_GET_SUBKEY = 39600;


    private static OpenPgpApi mApi;
    private static IOpenPgpService2 mService;
    private static ServiceConnection mServiceConnection;
    private static Context mCtx;

//    public static void onPackageStatusChanged(Context context, String action, String packagename) {
//        if (action==Intent.ACTION_PACKAGE_ADDED || action==Intent.ACTION_PACKAGE_CHANGED) {
//            if (PACKAGE_NAME.equals(packagename)) {
//                //seems like Open-Keychain just got installed or re-enabled
//                init(context);
//            }
//        }
//    }

    static {
        AppsReceiver.addListener(new AppsReceiver.IAppsReceiverListener() {
            @Override
            public void onAppChanged(Context ctx, String action, String packagename) {
                if (action.equals(Intent.ACTION_PACKAGE_ADDED) || action.equals(Intent.ACTION_PACKAGE_CHANGED)) {
                    if (PACKAGE_NAME.equals(packagename)) {
                        //seems like Open-Keychain just got installed or re-enabled
                        init(ctx);
                    }
                }
            }
        });
    }

    public static synchronized void init(final Context ctx) {
        if (mCtx == null) {
            mCtx = ctx.getApplicationContext();

            mServiceConnection = new ServiceConnection() {
                public void onServiceConnected(ComponentName name, IBinder service) {
                    Ln.w("OKCS: Service Connected!");
                    mService = IOpenPgpService2.Stub.asInterface(service);
                    mApi = new OpenPgpApi(mCtx, mService);
                }

                public void onServiceDisconnected(ComponentName name) {
                    Ln.w("OKCS: Service Disconnected!");
                    mService = null;
                    mApi = null;
                }
            };
        }


        bindToService(mCtx);
    }

    public static synchronized Intent executeApi(Intent data, InputStream is, OutputStream os) {
        Ln.w("OKCS: executeAPI... %s", data.getAction());
        if (mApi == null) {

            Intent res = new Intent();
            res.putExtra(OpenPgpApi.RESULT_CODE, OpenPgpApi.RESULT_CODE_ERROR);
            OpenPgpError error = new OpenPgpError(0, "OKC Service Not Bound, try again!!");
            res.putExtra(OpenPgpApi.RESULT_ERROR, error);

            //try to restart
            if (mCtx != null) {
                bindToService(mCtx);
            }

            return res;
        }

        Intent res = mApi.executeApi(data, is, os);
        Ln.w("OKCS: ...executeAPI");
        return res;
    }


    private static synchronized void bindToService(Context ctx) {
        Ln.w("OKCS: bindToService...");
        // if not already bound...
        if (mService == null) {


            if (getVersion(ctx) < V_MIN) {
                Ln.w("OKCS: not trying to connect to unsupported Open-Keychain Version %s", getVersion(ctx));
            } else {


                try {

                    Intent serviceIntent = new Intent(OpenPgpApi.SERVICE_INTENT_2);
                    // NOTE: setPackage is very important to restrict the intent to this provider only!
                    serviceIntent.setPackage(PACKAGE_NAME);
                    boolean connect = ctx.bindService(serviceIntent, mServiceConnection,
                            Context.BIND_AUTO_CREATE);
                    if (!connect) {
                        Ln.w("OKCS: Service couldn't be connected to, connect=false!");
                    }
                } catch (Exception e) {
                    Ln.e(e, "OKCS: Service couldn't be connected to");
                }
            }
        } else {
            Ln.w("OKCS: Service already/still bound");
        }
    }

    private static synchronized void unbindFromService() {
        if (mCtx != null && mServiceConnection != null) {
            mCtx.unbindService(mServiceConnection);
        }
    }

    public static boolean isInstalled(Context ctx) {
        return getVersion(ctx) >= 0;
    }

    public static int getVersion(Context context) {
        Intent intent = new Intent(OpenPgpApi.SERVICE_INTENT_2);
        intent.setPackage(PACKAGE_NAME);
        List<ResolveInfo> resInfo = context.getPackageManager().queryIntentServices(intent, 0);
        int res = -1;
        if (resInfo != null && !resInfo.isEmpty()) {
            ResolveInfo r = resInfo.get(0);
            ServiceInfo si = r.serviceInfo;
            if (si != null) {
                String pn = si.packageName;
                try {
                    res = context.getPackageManager()
                            .getPackageInfo(pn, 0).versionCode;
                } catch (PackageManager.NameNotFoundException e) {
                    e.printStackTrace();
                }
            }
        }
        return res;
    }

    public static String getVersionName(Context context) {
        Intent intent = new Intent(OpenPgpApi.SERVICE_INTENT_2);
        intent.setPackage(PACKAGE_NAME);
        List<ResolveInfo> resInfo = context.getPackageManager().queryIntentServices(intent, 0);
        String res = null;
        if (!resInfo.isEmpty()) {
            ResolveInfo r = resInfo.get(0);
            ServiceInfo si = r.serviceInfo;
            if (si != null) {
                String pn = si.packageName;
                try {
                    res = context.getPackageManager()
                            .getPackageInfo(pn, 0).versionName;
                } catch (PackageManager.NameNotFoundException e) {
                    e.printStackTrace();
                }
            }
        }
        return res;
    }

    public static Set<String> allPackageNames() {
        HashSet<String> res = new HashSet<>();
        res.add(PACKAGE_NAME);
        res.add("org.sufficientlysecure.keychain.debug");
        return res;
    }

    public static void doPanicExit(Context ctx) {
        try {
            Intent panicButtonIntent = new Intent("info.guardianproject.panic.action.TRIGGER");
            panicButtonIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            panicButtonIntent.setPackage(PACKAGE_NAME);
            ctx.startActivity(panicButtonIntent);
        } catch (Exception ex) {
            //ignore, probably not installed
        }
    }

    public static boolean isGooglePlayInstalled(Context context) {
        PackageManager pm = context.getPackageManager();
        try {
            pm.getPackageInfo("com.android.vending", 0);
            return true;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    public static void openInPlayStore(Activity activity) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + PACKAGE_NAME));
            activity.startActivity(intent);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public static void openInFdroid(Activity activity) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("http://www.f-droid.org/app/" + PACKAGE_NAME));
            activity.startActivity(intent);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }


}

File: crypto/src/main/java/io/oversec/one/crypto/gpg/OpenKeychainConnector.kt
package io.oversec.one.crypto.gpg

import android.annotation.SuppressLint
import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.net.Uri
import android.os.IBinder
import io.oversec.one.crypto.AppsReceiver
import org.openintents.openpgp.IOpenPgpService2
import org.openintents.openpgp.OpenPgpError
import org.openintents.openpgp.util.OpenPgpApi
import roboguice.util.Ln
import java.io.InputStream
import java.io.OutputStream
import java.util.HashSet


class OpenKeychainConnector(val ctx: Context) {

    private var mApi: OpenPgpApi? = null
    private var mService: IOpenPgpService2? = null
    private var mServiceConnection: ServiceConnection? = null


    //    public static void onPackageStatusChanged(Context context, String action, String packagename) {
    //        if (action==Intent.ACTION_PACKAGE_ADDED || action==Intent.ACTION_PACKAGE_CHANGED) {
    //            if (PACKAGE_NAME.equals(packagename)) {
    //                //seems like Open-Keychain just got installed or re-enabled
    //                init(context);
    //            }
    //        }
    //    }

    init {
        AppsReceiver.addListener(object : AppsReceiver.IAppsReceiverListener {
            override fun onAppChanged(ctx: Context, action: String, packagename: String) {
                if (action == Intent.ACTION_PACKAGE_ADDED || action == Intent.ACTION_PACKAGE_CHANGED) {
                    if (PACKAGE_NAME == packagename) {
                        //seems like Open-Keychain just got installed or re-enabled
                        init()
                    }
                }
            }
        })
    }


    fun init() {
        mServiceConnection = object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName, service: IBinder) {
                Ln.w("OKCS: Service Connected!")
                mService = IOpenPgpService2.Stub.asInterface(service)
                mApi = OpenPgpApi(ctx, mService)
            }

            override fun onServiceDisconnected(name: ComponentName) {
                Ln.w("OKCS: Service Disconnected!")
                mService = null
                mApi = null
            }
        }

        bindToService()
    }

    @Synchronized
    fun executeApi(data: Intent, inputStream: InputStream?, outputStream: OutputStream?): Intent {
        Ln.w("OKCS: executeAPI... %s", data.action)
        if (mApi == null) {

            val res = Intent()
            res.putExtra(OpenPgpApi.RESULT_CODE, OpenPgpApi.RESULT_CODE_ERROR)
            val error = OpenPgpError(0, "OKC Service Not Bound, try again!!")
            res.putExtra(OpenPgpApi.RESULT_ERROR, error)

            //try to restart
            bindToService()

            return res
        }
        else {
            val res = mApi!!.executeApi(data, inputStream, outputStream)
            Ln.w("OKCS: ...executeAPI")
            return res
        }
    }


    @Synchronized
    private fun bindToService() {
        Ln.d(Exception(), "DAMNIT OKCS: bindToService...")
        // if not already bound...
        if (mService == null) {


            if (getVersion() < V_MIN) {
                Ln.w(
                    "OKCS: not trying to connect to unsupported Open-Keychain Version %s",
                    getVersion()
                )
            } else {


                try {

                    val serviceIntent = Intent(OpenPgpApi.SERVICE_INTENT_2)
                    // NOTE: setPackage is very important to restrict the intent to this provider only!
                    serviceIntent.setPackage(PACKAGE_NAME)
                    val connect = ctx.bindService(
                        serviceIntent, mServiceConnection!!,
                        Context.BIND_AUTO_CREATE
                    )
                    if (!connect) {
                        Ln.w("OKCS: Service couldn't be connected to, connect=false!")
                    }
                } catch (e: Exception) {
                    Ln.e(e, "OKCS: Service couldn't be connected to")
                }

            }
        } else {
            Ln.w("OKCS: Service already/still bound")
        }
    }

    @Synchronized
    private fun unbindFromService() {
        if (mServiceConnection != null) {
            ctx.unbindService(mServiceConnection!!)
        }
    }

    fun isInstalled(): Boolean {
        return getVersion() >= 0
    }

    fun getVersion(): Int {
        val intent = Intent(OpenPgpApi.SERVICE_INTENT_2)
        intent.setPackage(PACKAGE_NAME)
        val resInfo = ctx.packageManager.queryIntentServices(intent, 0)
        var res = -1
        if (resInfo != null && !resInfo.isEmpty()) {
            val r = resInfo[0]
            val si = r.serviceInfo
            if (si != null) {
                val pn = si.packageName
                try {
                    res = ctx.packageManager
                        .getPackageInfo(pn, 0).versionCode
                } catch (e: PackageManager.NameNotFoundException) {
                    e.printStackTrace()
                }

            }
        }
        return res
    }

    fun getVersionName(): String? {
        val intent = Intent(OpenPgpApi.SERVICE_INTENT_2)
        intent.setPackage(PACKAGE_NAME)
        val resInfo = ctx.packageManager.queryIntentServices(intent, 0)
        var res: String? = null
        if (!resInfo.isEmpty()) {
            val r = resInfo[0]
            val si = r.serviceInfo
            if (si != null) {
                val pn = si.packageName
                try {
                    res = ctx.packageManager
                        .getPackageInfo(pn, 0).versionName
                } catch (e: PackageManager.NameNotFoundException) {
                    e.printStackTrace()
                }

            }
        }
        return res
    }

    fun allPackageNames(): Set<String> {
        val res = HashSet<String>()
        res.add(PACKAGE_NAME)
        res.add("org.sufficientlysecure.keychain.debug")
        return res
    }

    fun doPanicExit() {
        try {
            val panicButtonIntent = Intent("info.guardianproject.panic.action.TRIGGER")
            panicButtonIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            panicButtonIntent.setPackage(PACKAGE_NAME)
            ctx.startActivity(panicButtonIntent)
        } catch (ex: Exception) {
            //ignore, probably not installed
        }

    }

    fun isGooglePlayInstalled(): Boolean {
        val pm = ctx.packageManager
        try {
            pm.getPackageInfo("com.android.vending", 0)
            return true
        } catch (e: PackageManager.NameNotFoundException) {
            return false
        }

    }

    fun openInPlayStore() {
        try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$PACKAGE_NAME"))
            ctx.startActivity(intent)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }

    }

    fun openInFdroid() {
        try {
            val intent =
                Intent(Intent.ACTION_VIEW, Uri.parse("http://www.f-droid.org/app/$PACKAGE_NAME"))
            ctx.startActivity(intent)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }

    }

    companion object {
        const val PACKAGE_NAME =
            "org.sufficientlysecure.keychain" //"org.sufficientlysecure.keychain.debug";

        const val V_MIN = 39500  //The minimum version we support, kind of arbitrary decision
        const val V_GET_SUBKEY = 39600

        @SuppressLint("StaticFieldLeak") // note that we're storing *Application*context
        @Volatile
        private var INSTANCE: OpenKeychainConnector? = null

        fun getInstance(ctx: Context): OpenKeychainConnector =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: OpenKeychainConnector(ctx.applicationContext).also { INSTANCE = it }.also { it.init() }
            }

    }

}
